OSSL_{HTTP_,}parse_url(): Handle any scheme, userinfo, query, and fragment components

DDvO opened the pull request:
Extend `OSSL_HTTP_parse_url()` to handle `[http[s]://][userinfo@]host[:port][/path[?query]][#frag]`
~by silently ignoring~ to optionally provide any `userinfo`, `query`, and `frag` components of its input.
All current usages of this function, which are client-only, silently ignore userinfo and frag components,
while the query component is taken as part of the path.

This came up in https://github.com/openssl/openssl/issues/14001#issuecomment-769163033.

Also introduce a generalization `OSSL_parse_url()`, which handles any scheme (not only `http[s]`).

On this occasion document and deprecate `OCSP_parse_url()`.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

richsalz commented:
Nice work!  Pretty quick, too.

levitte commented:
Hmmm...  I've a curve ball for you: could the stripped out part be of interest to the caller?  This is, after all, a generic function...  and it's *new*, so now's the time to add output arguments without worrying about backward compatibility.
(for extra points, separate out the query part as well, if there is one)

DDvO commented:
> Hmmm... I've a curve ball for you: could the stripped out part be of interest to the caller? This is, after all, a generic function... and it's _new_, so now's the time to add output arguments

> (for extra points, separate out the query part as well, if there is one)

All right, so I've added the capability to return also any username, query, and fragment components.
I've also updated the unit tests and the documentation of the affected apps and lib functions.
In total this was much more effort than expected. 

> without worrying about backward compatibility.

I had to worry - there was is still  `OCSP_parse_url()`, and on this occasion I documented and deprecated it.

levitte commented:
> I had to worry - there was is still  `OCSP_parse_url()`, and on this occasion I documented and deprecated it.

Different *name*, that makes `OSSL_HTTP_parse_url()` a *new function*.

levitte commented:
> In total this was much more effort than expected.

Believe you me, I can relate.  It seems like half of the "easy targets" I take on turn out to be rabbit holes :wink:

beldmit commented:
Just a half? You are lucky :)

levitte commented on doc/man3/OSSL_HTTP_parse_url.pod:
```diff
@@ -0,0 +1,74 @@
+
+OSSL_HTTP_parse_url,
+OCSP_parse_url
+- http utilitiy functions
```
s/utilitiy/utility/

DDvO replied:
Fixed

levitte commented:
It's looking a lot better.

DDvO commented:
> > In total this was much more effort than expected.
>
> Believe you me, I can relate. It seems like half of the "easy targets" I take on turn out to be rabbit holes wink

In this case the awkwardness of string handling (including memory management) in C was the main reason 
(Perl would have been fantastically convenient here),
but usually in my experience the extra trouble is stumbling over technical debt in the libs, apps, tests, and/or docs.

DDvO commented:
> It's looking a lot better.

Good to hear.
I've now updated and extended also the commit message.

DDvO commented:
Sigh, there are weird unrelated failures of `30-test_evp_extra.t` in the sanitizer CI build, and I cannot even reproduce them locally.

levitte commented:
> Sigh, there are weird unrelated failures

Agreed that they are unrelated, so they should not be a factor when reviewing this PR

levitte commented on crypto/http/http_lib.c:
```diff
@@ -112,14 +138,36 @@ int OSSL_HTTP_parse_url(const char *url, char **phost, char **pport,
+    if (pquery != NULL && tmp != NULL) {
+        path_end = tmp;
+        query = path_end + 1;
+        query_end = frag = frag_end = query + strlen(query);
```
Why do you think that `query + strlen(query)` will give a different result than `path + strlen(path)` on line 141?  It looks to me like this line is superfluous...

DDvO replied:
Good point - I had put the line for symmetry with the above, not realizing that it is redundant.
Removed, as well as the one below.

levitte commented on crypto/http/http_lib.c:
```diff
@@ -112,14 +138,36 @@ int OSSL_HTTP_parse_url(const char *url, char **phost, char **pport,
+            path_end = tmp;
+        query_end = tmp;
+        frag = query_end + 1;
+        frag_end = frag + strlen(frag);
```
Same thing here, `frag_end` is unlikely to get a different value than on line 141...

levitte commented on doc/man3/OSSL_HTTP_parse_url.pod:
```diff
@@ -0,0 +1,74 @@
+=head1 DESCRIPTION
+
+OSSL_HTTP_parse_url() parses its input string I<url> as a URL of the form
+C<[http[s]://][userinfo@]host[:port][/path[?query]][#fragment]> and splits it up
```
This is slightly incorrect from a standards standpoint, but what's more, this is not what I see your code doing.  To be perfectly correct (and what the function seems to do), that should be:

`C<[http[s]://][userinfo@]host[:port][/path][?query][#fragment]>`

In other words, the query part isn't *part of* the path.

richsalz replied:
> `C<[http[s]://][userinfo@]host[:port][/path][?query][#fragment]>`

It should be `/[path]` since there must be a slash after the hostpart. Or just `/path` and indicate that it can be an empty string.

I also think it's a mistake to make the protocol be optional. Even though the function has `http` in the name, you're allowing both http and https, and it's forward-compatible for when you add other URI types.

I also think, and maybe not in this manpage or PR, you need to make *very clear* that https is not supported. It should be an error if the client tries to use https.

DDvO replied:
Good catch!
I had stated the pattern this way in order to reflect that the client side can see the query component as part of the path.
This morning I thought that this pattern is not really correct, but then decided to leave it in for now, 
as a kind of Easter Egg to be discovered by a careful reviewer - and inevitably you found it :+1: 
Fixed.

richsalz replied:
glad you're fixing the pattern.  and the other two comments?

levitte commented on doc/man3/OSSL_HTTP_parse_url.pod:
```diff
@@ -0,0 +1,74 @@
+The port component is optional and defaults to "443" for HTTPS, else "80".
+If the I<pport_num> argument is NULL the port specification can be in mnemonic
+form such as C<http> like with L<BIO_set_conn_port(3)>, else it must be in
+numerical form and its integer value is assigned to I<*pport_num> on success.
```
https://tools.ietf.org/html/rfc3986#section-3.2.3

You're likely never to see a service name used as a port, as that's invalid by the standard.

DDvO replied:
True that this feature of `BIO_set_conn_port()` is hardly (known and) used,
and since moreover RFC 3986 forbids it for URLs I've disabled it already for the parsing.

DDvO commented:
> In this case the awkwardness of string handling (including memory management) in C was the main reason

While doing the fixups that came up during the review 
I found myself that some corner cases (such as empty path but non-empty query or fragment)
were not correctly handled yet. So I've just fixed them and added respective tests.

DDvO commented:
Having a look again at the code I felt that the code inherited from `OCSP_parse_url()` was still rather messy.
I've cleaned it up further and made it more efficient, 
in particular getting rid of the dynamically allocated buffer that was used to mess with a copy of the input `url`.
Now all string pointers are of type `const char *`.

DDvO commented:
Rebased on the latest master in order to merge related documentation changes.

DDvO commented:
Can this be approved now?

DDvO commented:
Thanks @levitte for approving #14074.
How about getting this PR done as well?

DDvO commented:
@levitte, how to bring this forward?

levitte commented on doc/man3/OSSL_HTTP_parse_url.pod:
```diff
@@ -0,0 +1,73 @@
+ int OSSL_HTTP_parse_url(const char *url,
+                         int *pssl, char **puser, char **phost,
+                         char **pport, int *pport_num,
+                         char **ppath, char **pquery, char **pfrag);
```
As a general parser, I can see one piece missing; I think it would be interesting for the caller to know the scheme too.  People who want to write OSSL_STORE backends that handle all sorts of schemes (whadoIknow, someone might come up with an LDAP implementation, for example) could find that useful.  In that light, I do question the general use of `*pssl`, as that's probably better handled by the caller, who should know better exactly which scheme they are interested in.

DDvO replied:
All right, done.
This was again much more effort than expected dues to the nastiness of C strings.

levitte replied:
> dues to the nastiness of C strings.

Eh gods, tell me about it.

levitte commented:
The CIs are complaining:

```
crypto/http/http_lib.c: In function 'OSSL_parse_url':
crypto/http/http_lib.c:114:50: error: comparison of unsigned expression < 0 is always false [-Werror=type-limits]
     if (!sscanf(port, "%u", &portnum) || portnum < 0 || portnum > 65535) {
                                                  ^
cc1: all warnings being treated as errors
```

and true, `%u` will never scan into a negative number, so it should be a matter of just removing `|| portnum < 0`

DDvO commented:
> and true, `%u` will never scan into a negative number, so it should be a matter of just removing `|| portnum < 0`

Yep, fixed.

openssl-machine commented:
This pull request is ready to merge

DDvO commented:
Oops, this PR could have been merged since more than 11 days - done now.

Thanks @levitte for your review including good suggestions on generalizations.
